vapi/library/finder/path.go

```diff
@@ -135,79 +135,125 @@ func (f *PathFinder) datastoreName(ctx context.Context, id string) (string, erro
 	return name, nil
 }
 
-func (f *PathFinder) convertPath(ctx context.Context, b *mo.Datastore, path string) (string, error) {
-	if !internal.IsDatastoreVSAN(*b) {
+func (f *PathFinder) convertPath(
+	ctx context.Context,
+	dc *object.Datacenter,
+	ds mo.Datastore,
+	path string) (string, error) {
+
+	if v := ds.Capability.TopLevelDirectoryCreateSupported; v != nil && *v {
 		return path, nil
 	}
 
-	var dc *object.Datacenter
-
-	entities, err := mo.Ancestors(ctx, f.c, f.c.ServiceContent.PropertyCollector, b.Self)
-	if err != nil {
-		return "", err
+	if dc == nil {
+		entities, err := mo.Ancestors(
+			ctx,
+			f.c,
+			f.c.ServiceContent.PropertyCollector,
+			ds.Self)
+		if err != nil {
+			return "", fmt.Errorf("failed to find ancestors: %w", err)
+		}
+		for _, entity := range entities {
+			if entity.Self.Type == "Datacenter" {
+				dc = object.NewDatacenter(f.c, entity.Self)
+				break
+			}
+		}
 	}
 
-	for _, entity := range entities {
-		if entity.Self.Type == "Datacenter" {
-			dc = object.NewDatacenter(f.c, entity.Self)
-			break
-		}
+	if dc == nil {
+		return "", errors.New("failed to find datacenter")
 	}
 
 	m := object.NewDatastoreNamespaceManager(f.c)
 	return m.ConvertNamespacePathToUuidPath(ctx, dc, path)
 }
 
-// ResolveLibraryItemStorage transforms StorageURIs Datastore url (uuid) to Datastore name.
-func (f *PathFinder) ResolveLibraryItemStorage(ctx context.Context, storage []library.Storage) error {
+// ResolveLibraryItemStorage calls ResolveLibraryItemStorageWithDatacenter with
```

Comment:
I think it'd be ok to break the `ResolveLibraryItemStorage` signature and add the Datacenter param.
We recently added it, only use other than govc is image-registry-operator